README.md
# storybook-addon-react-docgen

[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=for-the-badge)](https://github.com/prettier/prettier) [![CircleCI](https://img.shields.io/circleci/project/github/hipstersmoothie/storybook-addon-react-docgen/master.svg?style=for-the-badge)](https://circleci.com/gh/hipstersmoothie/storybook-addon-react-docgen) [![npm](https://img.shields.io/npm/v/storybook-addon-react-docgen.svg?style=for-the-badge)](https://www.npmjs.com/package/storybook-addon-react-docgen) [![npm](https://img.shields.io/npm/dt/storybook-addon-react-docgen.svg?style=for-the-badge)](https://www.npmjs.com/package/storybook-addon-react-docgen)

A storybook addon to display react docgen info.
This addon is a drop in replacement for the "info" addon's prop table functionality.
Rather than rendering with the component it renders in the addons panel.
Works with typescript too!

There exist other addons that do this, but they didn't work in the same way as the `info` addon.
This resulted in complicated configuration changes.
This plugin aims to be painless to switch to.

![Example Output](https://github.com/hipstersmoothie/storybook-addon-react-docgen/raw/master/example.png)

## Installation

```sh
yarn add storybook-addon-react-docgen
```

## React Docgen Integration

React Docgen is included as part of the `@storybook/addon-docs` package.
If you are using `@storybook/addon-docs` then you do not need to set up docgen and can skip the next steps

### Typescript DocGen

To use this plugin with a typescript project you need to install [react-docgen-typescript-plugin](https://www.npmjs.com/package/react-docgen-typescript-plugin) and configure webpack to use it.

### Javascript DocGen

To use this plugin with a javascript project you need to install [babel-plugin-react-docgen](https://github.com/storybookjs/babel-plugin-react-docgen)

## Usage

Add it in your `main.js` addons":

```js
module.exports = {
  stories: ['../stories/**/*.stories.js'],
  addons: ['@storybook/addon-docs', 'storybook-addon-react-docgen']
};
```

Then add the `withPropsTable` decorator to your `preview.js`.
You can pass global options here if you want:

```js
const { addDecorator } = require('@storybook/react');
const { withPropsTable } = require('storybook-addon-react-docgen');

addDecorator(withPropsTable(options));
// or
addDecorator(withPropsTable);
```

You can use the `props` parameter to configure the options for individual stories:

```js
import { storiesOf } from '@storybook/react';

import Other from './Other';
import Component from './Component';

export default {
  title: 'Components/Button'
};

export const WithSomeEmoji = () => (
  <Component>
    <Other />
  </Component>
);

WithSomeEmoji.parameters: {
 props: {
   propTablesExclude: [
     Other // the actual component
   ]
 }
}
```

or for the entire story:

```js
import { storiesOf } from '@storybook/react';

import Other from './Other';
import Component from './Component';

export default {
  title: 'Components/Button',
  parameters: {
    props: {
      propTablesExclude: [
        Other // the actual component
      ]
    }
  }
};

export const WithSomeEmoji = () => (
  <Component>
    <Other />
  </Component>
);
```

## Configuration

```js
{
  /**
   * Components used in story
   * Displays Prop Tables with these components
   * @default []
   */
  propTables: Array<React.ComponentType>,

  /**
   * Define custom sorting order for the components specifying component names in the desired order.
   * Example:
   * propTablesSortOrder: ["MyComponent", "FooComponent", "AnotherComponent"]
   * @default []
   */
  propTablesSortOrder: string,
  /**
   * Only include prop tables for these components.
   * Accepts an array of component classes or functions
   * @default null
   */
  propTablesInclude: Array<React.ComponentType | string>,
  /**
   * Exclude Components from being shown in Prop Tables section
   * Accepts an array of component classes or functions
   * @default []
   */
  propTablesExclude: Array<React.ComponentType | string>,
  /**
   * Overrides styles of addon. The object should follow this shape:
   * https://github.com/storybooks/storybook/blob/master/addons/info/src/components/Story.js#L19.
   * This prop can also accept a function which has the default stylesheet passed as an argument
   */
  styles: Object | Function,
  /**
   * Max props to display per line in source code
   * @default 3
   */
  maxPropsIntoLine: number,
  /**
   * Displays the first 10 characters of the prop name
   * @default 3
   */
  maxPropObjectKeys: number,
  /**
   * Displays the first 10 items in the default prop array
   * @default 3
   */
  maxPropArrayLength: number,
  /**
   * Displays the first 100 characters in the default prop string
   * @default 50
   */
  maxPropStringLength: number,
  /**
   * Override the component used to render the props table
   * @default PropTable
   */
  TableComponent: React.ComponentType,
  /**
   * Will exclude any respective properties whose name is included in array.
   * Can also specify absolute propType to exclude (see example below)
   * Examples:
   * excludedPropTypes: ["message"] // propType to exclude
   * excludedPropTypes: ["MyComponent.message"] // absolute propType
   *
   * @default []
   */
  excludedPropTypes: Array<string>,
}
```

### Rendering a Custom Table

The TableComponent option allows you to define how the prop table should be rendered. Your component will be rendered with the following props.

```js
{
  propDefinitions: Array<{
    property: string, // The name of the prop
    propType: Object | string, // The prop type. TODO: info about what this object is...
    required: boolean, // True if the prop is required
    description: string, // The description of the prop
    defaultValue: any // The default value of the prop
  }>
}
```

## FAQ

### Nothing shows up, this is broken!

The way that the packages implement docgen for react means that there are some limitations on how you can import things and write you components.

1. Must use default export + named export: The docgen will not be able to pick up a name for the default export so you must also use a named export

```ts
import * as React from "react";

interface ColorButtonProps {
  /** Buttons background color */
  color: "blue" | "green";
}

/** A button with a configurable background color. */
export const ColorButton: React.SFC<ColorButtonProps> = props => (
  <button {...props}>
);

export default ColorButton;
```

2. Imports Matter (TypeScript only): The way you import react and use it's types must conform to a few different formate

```tsx
// To get React.FC to work
import * as React from 'react';
export const Button: React.FC<ButtonProps> = () => {};

// Without "* as" you can only use like:
import React, { FC } from 'react';
export const Button: FC<ButtonProps> = () => {};
```

3. Usage within the story matter: This addon determines what components to display props for by finding all components used in the JSX returned by the story. So if you want prop-types to be displayed for a component, you **must** return that component in the story function.

```tsx
import React from 'react';
import { Button } from './Button';

export default {
  title: 'Button'
};

/** WILL NOT WORK */

// Since the usage of the component is not in the JSX
// returned by the story function no props are displayed
const MyFancyExample = () => {
  const [count, setCount] = React.useState(0);

  return (
    <Button
      primary={boolean('primary', false)}
      onClick={() => setCount(count + 1)}
    >
      "hello: " {count}
    </Button>
  );
};

export const BaseStory = () => <MyFancyExample />;

/** WILL WORK */

// The JSX returned by the story uses Button, so we will
// get the props types for button.
export const BaseStory = () => {
  const [count, setCount] = React.useState(0);

  return (
    <Button
      primary={boolean('primary', false)}
      onClick={() => setCount(count + 1)}
    >
      "hello: " {count}
    </Button>
  );
};
```

### Why are default props so hard to get right? (TypeScript only)

The `react` types are magical and you're probably doing too much. Using `React.FC` is the quickest way to ramp up the complexity of your components. Once you use that you lose the `defaultProps` experience.

**Using `React.FC`:**

```tsx
interface CardProps {
  size?: 'small' | 'large';
}

// The type of size will be "string | undefined"
// You will either have to repeat your default value
// Or write a helper type the figures out what is defaulted
const Card: React.FC<CardProps> = ({ size }) => <div>{size}</div>;

Card.defaultProps = {
  size: 'small'
};

// Size is optional to the user
const Usage = () => <Card />;
```

Without React.FC:

```tsx
interface CardProps {
  // Key part right here is to make the defaulted prop not optional
  // this way in your function it won't be undefined
  size: 'small' | 'large';
}

// The type of size will be "string"
const Card = ({ size }: CardProps) => <div>{size}</div>;

// Typescript can use this defaultProps to determine what is optional
// for the user of your component.
Card.defaultProps = {
  size: 'small'
};

// Size is optional to the user
const Usage = () => <Card />;
```

### My components extends from HTML elements and there are way too many props in the panel! How do I get rid of some?

You can add a filter to `react-docgen-typescript-plugin` that will omit anything that comes from `@types/react`.

`.storybook/main.js`

```js
{
    typescript: {
        check: false,
        reactDocgen: 'react-docgen-typescript',
        reactDocgenTypescriptOptions: {
            propFilter: (prop) => {
                if (prop.name === 'children') {
                    return true;
                }

                if (prop.parent) {
                    return (
                        !/@types\/react/.test(prop.parent.fileName) &&
                        !/@emotion/.test(prop.parent.fileName)
                    );
                }
                return true;
            },
        },
    },
}
```

## Inspiration

Code heavily inspired by (copied from):

- [@storybook/addon-info](https://www.npmjs.com/package/@storybook/addon-info)
- [@storybook/addon-notes](https://www.npmjs.com/package/@storybook/addon-notes)


CHANGELOG.md
# v1.2.44 (Tue Aug 22 2023)

:tada: This release contains work from a new contributor! :tada:

Thank you, Yucho Ho ([@yucho](https://github.com/yucho)), for all your work!

#### 🐛 Bug Fix

- fix: remove unused deprecated package [#140](https://github.com/hipstersmoothie/storybook-addon-react-docgen/pull/140) ([@yucho](https://github.com/yucho))

#### Authors: 1

- Yucho Ho ([@yucho](https://github.com/yucho))

---

# v1.2.43 (Mon Feb 06 2023)

#### 🐛 Bug Fix

- Update setup instructions [#138](https://github.com/hipstersmoothie/storybook-addon-react-docgen/pull/138) ([@hipstersmoothie](https://github.com/hipstersmoothie))

#### Authors: 1

- Andrew Lisowski ([@hipstersmoothie](https://github.com/hipstersmoothie))

---

# v1.2.42 (Fri Jul 17 2020)

#### 🐛 Bug Fix

- Needed to make revisions to prevent build from breaking. [#114](https://github.com/hipstersmoothie/storybook-addon-react-docgen/pull/114) ([@sandrews2](https://github.com/sandrews2))
- Change type checking of object to use Object.is [#113](https://github.com/hipstersmoothie/storybook-addon-react-docgen/pull/113) ([@sandrews2](https://github.com/sandrews2))

#### Authors: 1

- [@sandrews2](https://github.com/sandrews2)

---

# v1.2.41 (Tue Jul 07 2020)

#### 🐛 Bug Fix

- handle component supplied as props [#112](https://github.com/hipstersmoothie/storybook-addon-react-docgen/pull/112) ([@hipstersmoothie](https://github.com/hipstersmoothie))

#### Authors: 1

- Andrew Lisowski ([@hipstersmoothie](https://github.com/hipstersmoothie))

---

# v1.2.40 (Tue Jun 23 2020)

#### 🐛 Bug Fix

- improve complex object default prop rendering [#111](https://github.com/hipstersmoothie/storybook-addon-react-docgen/pull/111) ([@hipstersmoothie](https://github.com/hipstersmoothie))

#### Authors: 1

- Andrew Lisowski ([@hipstersmoothie](https://github.com/hipstersmoothie))

---

# v1.2.39 (Wed Jun 17 2020)

#### 🐛 Bug Fix

- Typescript Rewrite [#110](https://github.com/hipstersmoothie/storybook-addon-react-docgen/pull/110) ([@hipstersmoothie](https://github.com/hipstersmoothie))

#### Authors: 1

- Andrew Lisowski ([@hipstersmoothie](https://github.com/hipstersmoothie))

---

# v1.2.38 (Wed Jun 10 2020)

#### 🐛 Bug Fix

- chore: filter React Fragments [#106](https://github.com/hipstersmoothie/storybook-addon-react-docgen/pull/106) ([@petermikitsh](https://github.com/petermikitsh))

#### Authors: 1

- Peter Mikitsh ([@petermikitsh](https://github.com/petermikitsh))

---

# v1.2.37 (Tue Jun 09 2020)

#### 🐛 Bug Fix

- fix: switching between stories exception [#105](https://github.com/hipstersmoothie/storybook-addon-react-docgen/pull/105) ([@petermikitsh](https://github.com/petermikitsh))

#### Authors: 1

- Peter Mikitsh ([@petermikitsh](https://github.com/petermikitsh))

---

# v1.2.36 (Tue Jun 09 2020)

:tada: This release contains work from a new contributor! :tada:

Thank you, Peter Mikitsh ([@petermikitsh](https://github.com/petermikitsh)), for all your work!

#### 🐛 Bug Fix

- feat: interactive props panel [#104](https://github.com/hipstersmoothie/storybook-addon-react-docgen/pull/104) ([@petermikitsh](https://github.com/petermikitsh))

#### Authors: 1

- Peter Mikitsh ([@petermikitsh](https://github.com/petermikitsh))

---

# v1.2.35 (Mon Jun 08 2020)

#### 🐛 Bug Fix

- Use storybook-pretty-props [#102](https://github.com/hipstersmoothie/storybook-addon-react-docgen/pull/102) ([@hipstersmoothie](https://github.com/hipstersmoothie))

#### Authors: 1

- Andrew Lisowski ([@hipstersmoothie](https://github.com/hipstersmoothie))

---

# v1.2.34 (Fri Jun 05 2020)

#### 🐛  Bug Fix

- add "propTablesInclude" to include only the tables for those components [#101](https://github.com/hipstersmoothie/storybook-addon-react-docgen/pull/101) ([@hipstersmoothie](https://github.com/hipstersmoothie))

#### Authors: 1

- Andrew Lisowski ([@hipstersmoothie](https://github.com/hipstersmoothie))

---

# v1.2.33 (Thu May 28 2020)

#### 🐛  Bug Fix

- positional params error [#99](https://github.com/hipstersmoothie/storybook-addon-react-docgen/pull/99) ([@hainessss](https://github.com/hainessss))

#### Authors: 1

- [@hainessss](https://github.com/hainessss)

---

# v1.2.32 (Thu Apr 09 2020)

#### 🐛  Bug Fix

- handle when both name and displayname are undefined [#96](https://github.com/hipstersmoothie/storybook-addon-react-docgen/pull/96) ([@hipstersmoothie](https://github.com/hipstersmoothie))

#### Authors: 1

- Andrew Lisowski ([@hipstersmoothie](https://github.com/hipstersmoothie))

---

# v1.2.31 (Mon Apr 06 2020)

#### 🐛  Bug Fix

- feat: support propTablesSortOrder [#94](https://github.com/hipstersmoothie/storybook-addon-react-docgen/pull/94) ([@nischithbm](https://github.com/nischithbm))
- feat: support propTablesSortOrder [#1](https://github.com/hipstersmoothie/storybook-addon-react-docgen/pull/1) ([@hipstersmoothie](https://github.com/hipstersmoothie))

#### Authors: 2

- Nischith B.M. ([@nischithbm](https://github.com/nischithbm))
- Andrew Lisowski ([@hipstersmoothie](https://github.com/hipstersmoothie))

---

# v1.2.30 (Sun Apr 05 2020)

#### 🐛  Bug Fix

- fix: propTablesExclude to work for forwardRef(FC) [#90](https://github.com/hipstersmoothie/storybook-addon-react-docgen/pull/90) ([@nischithbm](https://github.com/nischithbm))

#### Authors: 1

- Nischith B.M. ([@nischithbm](https://github.com/nischithbm))

---

# v1.2.29 (Sun Apr 05 2020)

#### 🐛  Bug Fix

- feat: absolute propTypes in excludedPropTypes [#91](https://github.com/hipstersmoothie/storybook-addon-react-docgen/pull/91) ([@nischithbm](https://github.com/nischithbm))

#### 📝  Documentation

- Update README.md [#79](https://github.com/hipstersmoothie/storybook-addon-react-docgen/pull/79) ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Add usage docs [#78](https://github.com/hipstersmoothie/storybook-addon-react-docgen/pull/78) ([@hipstersmoothie](https://github.com/hipstersmoothie))

stories/button.js
import React from 'react';
import PropTypes from 'prop-types';

export const Button = props => <button {...props} />;

Button.propTypes = {
  /** A action for the button to take */
  action: PropTypes.string,
  /** A made-up property for small enum test case */
  smallEnum: PropTypes.oneOf(['a', 'b', 'c', 'd', 'e', 'f']),
  /** A made-up property to test minimizing/expanding enum list */
  mediumEnum: PropTypes.oneOf([
    'a',
    'b',
    'c',
    'd',
    'e',
    'f',
    'g',
    'h',
    'i',
    'j',
    'k',
    'l',
    'm',
    'n',
    'o'
  ]),
  /** A made-up property for overflow enum test case */
  largeEnum: PropTypes.oneOf([
    '0',
    '1',
    '2',
    '3',
    '4',
    '5',
    '6',
    '7',
    '8',
    '9',
    '10',
    '11',
    '12',
    '13',
    '14',
    '15',
    '16',
    '17',
    '18',
    '19',
    '20',
    '21',
    '22',
    '23',
    '24',
    '25',
    '26',
    '27',
    '28',
    '29',
    '30',
    '31',
    '32',
    '33',
    '34',
    '35',
    '36',
    '37',
    '38',
    '39',
    '40',
    '41',
    '42',
    '43',
    '44',
    '45',
    '46',
    '47',
    '48',
    '49',
    '50',
    '51',
    '52',
    '53',
    '54',
    '55',
    '56',
    '57',
    '58',
    '59',
    '60',
    '61',
    '62',
    '63',
    '64',
    '65',
    '66',
    '67',
    '68',
    '69',
    '70',
    '71',
    '72',
    '73',
    '74',
    '75',
    '76',
    '77',
    '78',
    '79',
    '80',
    '81',
    '82',
    '83',
    '84',
    '85',
    '86',
    '87',
    '88',
    '89',
    '90',
    '91',
    '92',
    '93',
    '94',
    '95',
    '96',
    '97',
    '98',
    '99',
    '100',
    '101',
    '102',
    '103',
    '104',
    '105',
    '106',
    '107',
    '108',
    '109',
    '110',
    '111',
    '112',
    '113',
    '114',
    '115',
    '116',
    '117',
    '118',
    '119',
    '120',
    '121',
    '122',
    '123',
    '124',
    '125',
    '126',
    '127',
    '128',
    '129',
    '130',
    '131',
    '132',
    '133',
    '134',
    '135',
    '136',
    '137',
    '138',
    '139',
    '140',
    '141',
    '142',
    '143',
    '144',
    '145',
    '146',
    '147',
    '148',
    '149',
    '150',
    '151',
    '152',
    '153',
    '154',
    '155',
    '156',
    '157',
    '158',
    '159',
    '160',
    '161',
    '162',
    '163',
    '164',
    '165',
    '166',
    '167',
    '168',
    '169',
    '170',
    '171',
    '172',
    '173',
    '174',
    '175',
    '176',
    '177',
    '178',
    '179',
    '180',
    '181',
    '182',
    '183',
    '184',
    '185',
    '186',
    '187',
    '188',
    '189',
    '190',
    '191',
    '192',
    '193',
    '194',
    '195',
    '196',
    '197',
    '198',
    '199',
    '200',
    '201',
    '202',
    '203',
    '204',
    '205',
    '206',
    '207',
    '208',
    '209',
    '210',
    '211',
    '212',
    '213',
    '214',
    '215',
    '216',
    '217',
    '218',
    '219',
    '220',
    '221',
    '222',
    '223',
    '224',
    '225',
    '226',
    '227',
    '228',
    '229',
    '230',
    '231',
    '232',
    '233',
    '234',
    '235',
    '236',
    '237',
    '238',
    '239',
    '240',
    '241',
    '242',
    '243',
    '244',
    '245',
    '246',
    '247',
    '248',
    '249',
    '250',
    '251',
    '252',
    '253',
    '254',
    '255',
    '256',
    '257',
    '258',
    '259',
    '260',
    '261',
    '262',
    '263',
    '264',
    '265',
    '266',
    '267',
    '268',
    '269',
    '270',
    '271',
    '272',
    '273',
    '274',
    '275',
    '276',
    '277',
    '278',
    '279',
    '280',
    '281',
    '282',
    '283',
    '284',
    '285',
    '286',
    '287',
    '288',
    '289',
    '290',
    '291',
    '292',
    '293',
    '294',
    '295',
    '296',
    '297',
    '298',
    '299',
    '300',
    '301',
    '302',
    '303',
    '304',
    '305',
    '306',
    '307',
    '308',
    '309',
    '310',
    '311',
    '312',
    '313',
    '314',
    '315',
    '316',
    '317',
    '318',
    '319',
    '320',
    '321',
    '322',
    '323',
    '324',
    '325',
    '326',
    '327',
    '328',
    '329',
    '330',
    '331',
    '332',
    '333',
    '334',
    '335',
    '336',
    '337',
    '338',
    '339',
    '340',
    '341',
    '342',
    '343',
    '344',
    '345',
    '346',
    '347',
    '348',
    '349',
    '350',
    '351',
    '352',
    '353',
    '354',
    '355',
    '356',
    '357',
    '358',
    '359',
    '360',
    '361',
    '362',
    '363',
    '364',
    '365',
    '366',
    '367',
    '368',
    '369',
    '370',
    '371',
    '372',
    '373',
    '374',
    '375',
    '376',
    '377',
    '378',
    '379',
    '380',
    '381',
    '382',
    '383',
    '384',
    '385',
    '386',
    '387',
    '388',
    '389',
    '390',
    '391',
    '392',
    '393',
    '394',
    '395',
    '396',
    '397',
    '398',
    '399',
    '400',
    '401',
    '402',
    '403',
    '404',
    '405',
    '406',
    '407',
    '408',
    '409',
    '410',
    '411',
    '412',
    '413',
    '414',
    '415',
    '416',
    '417',
    '418',
    '419',
    '420',
    '421',
    '422',
    '423',
    '424',
    '425',
    '426',
    '427',
    '428',
    '429',
    '430',
    '431',
    '432',
    '433',
    '434',
    '435',
    '436',
    '437',
    '438',
    '439',
    '440',
    '441',
    '442',
    '443',
    '444',
    '445',
    '446',
    '447',
    '448',
    '449',
    '450',
    '451',
    '452',
    '453',
    '454',
    '455',
    '456',
    '457',
    '458',
    '459',
    '460',
    '461',
    '462',
    '463',
    '464',
    '465',
    '466',
    '467',
    '468',
    '469',
    '470',
    '471',
    '472',
    '473',
    '474',
    '475',
    '476',
    '477',
    '478',
    '479',
    '480',
    '481',
    '482',
    '483',
    '484',
    '485',
    '486',
    '487',
    '488',
    '489',
    '490',
    '491',
    '492',
    '493',
    '494',
    '495',
    '496',
    '497',
    '498',
    '499',
    '500',
    '501',
    '502',
    '503',
    '504',
    '505',
    '506',
    '507',
    '508',
    '509',
    '510',
    '511',
    '512',
    '513',
    '514',
    '515',
    '516',
    '517',
    '518',
    '519',
    '520',
    '521',
    '522',
    '523',
    '524',
    '525',
    '526',
    '527',
    '528',
    '529',
    '530',
    '531',
    '532',
    '533',
    '534',
    '535',
    '536',
    '537',
    '538',
    '539',
    '540',
    '541',
    '542',
    '543',
    '544',
    '545',
    '546',
    '547',
    '548',
    '549',
    '550',
    '551',
    '552',
    '553',
    '554',
    '555',
    '556',
    '557',
    '558',
    '559',
    '560',
    '561',
    '562',
    '563',
    '564',
    '565',
    '566',
    '567',
    '568',
    '569',
    '570',
    '571',
    '572',
    '573',
    '574',
    '575',
    '576',
    '577',
    '578',
    '579',
    '580',
    '581',
    '582',
    '583',
    '584',
    '585',
    '586',
    '587',
    '588',
    '589',
    '590',
    '591',
    '592',
    '593',
    '594',
    '595',
    '596',
    '597',
    '598',
    '599',
    '600',
    '601',
    '602',
    '603',
    '604',
    '605',
    '606',
    '607',
    '608',
    '609',
    '610',
    '611',
    '612',
    '613',
    '614',
    '615',
    '616',
    '617',
    '618',
    '619',
    '620',
    '621',
    '622',
    '623',
    '624',
    '625',
    '626',
    '627',
    '628',
    '629',
    '630',
    '631',
    '632',
    '633',
    '634',
    '635',
    '636',
    '637',
    '638',
    '639',
    '640',
    '641',
    '642',
    '643',
    '644',
    '645',
    '646',
    '647',
    '648',
    '649',
    '650',
    '651',
    '652',
    '653',
    '654',
    '655',
    '656',
    '657',
    '658',
    '659',
    '660',
    '661',
    '662',
    '663',
    '664',
    '665',
    '666',
    '667',
    '668',
    '669',
    '670',
    '671',
    '672',
    '673',
    '674',
    '675',
    '676',
    '677',
    '678',
    '679',
    '680',
    '681',
    '682',
    '683',
    '684',
    '685',
    '686',
    '687',
    '688',
    '689',
    '690',
    '691',
    '692',
    '693',
    '694',
    '695',
    '696',
    '697',
    '698',
    '699',
    '700',
    '701',
    '702',
    '703',
    '704',
    '705',
    '706',
    '707',
    '708',
    '709',
    '710',
    '711',
    '712',
    '713',
    '714',
    '715',
    '716',
    '717',
    '718',
    '719',
    '720',
    '721',
    '722',
    '723',
    '724',
    '725',
    '726',
    '727',
    '728',
    '729',
    '730',
    '731',
    '732',
    '733',
    '734',
    '735',
    '736',
    '737',
    '738',
    '739',
    '740',
    '741',
    '742',
    '743',
    '744',
    '745',
    '746',
    '747',
    '748',
    '749',
    '750',
    '751',
    '752',
    '753',
    '754',
    '755',
    '756',
    '757',
    '758',
    '759',
    '760',
    '761',
    '762',
    '763',
    '764',
    '765',
    '766',
    '767',
    '768',
    '769',
    '770',
    '771',
    '772',
    '773',
    '774',
    '775',
    '776',
    '777',
    '778',
    '779',
    '780',
    '781',
    '782',
    '783',
    '784',
    '785',
    '786',
    '787',
    '788',
    '789',
    '790',
    '791',
    '792',
    '793',
    '794',
    '795',
    '796',
    '797',
    '798',
    '799',
    '800',
    '801',
    '802',
    '803',
    '804',
    '805',
    '806',
    '807',
    '808',
    '809',
    '810',
    '811',
    '812',
    '813',
    '814',
    '815',
    '816',
    '817',
    '818',
    '819',
    '820',
    '821',
    '822',
    '823',
    '824',
    '825',
    '826',
    '827',
    '828',
    '829',
    '830',
    '831',
    '832',
    '833',
    '834',
    '835',
    '836',
    '837',
    '838',
    '839',
    '840',
    '841',
    '842',
    '843',
    '844',
    '845',
    '846',
    '847',
    '848',
    '849',
    '850',
    '851',
    '852',
    '853',
    '854',
    '855',
    '856',
    '857',
    '858',
    '859',
    '860',
    '861',
    '862',
    '863',
    '864',
    '865',
    '866',
    '867',
    '868',
    '869',
    '870',
    '871',
    '872',
    '873',
    '874',
    '875',
    '876',
    '877',
    '878',
    '879',
    '880',
    '881',
    '882',
    '883',
    '884',
    '885',
    '886',
    '887',
    '888',
    '889',
    '890',
    '891',
    '892',
    '893',
    '894',
    '895',
    '896',
    '897',
    '898',
    '899',
    '900',
    '901',
    '902',
    '903',
    '904',
    '905',
    '906',
    '907',
    '908',
    '909',
    '910',
    '911',
    '912',
    '913',
    '914',
    '915',
    '916',
    '917',
    '918',
    '919',
    '920',
    '921',
    '922',
    '923',
    '924',
    '925',
    '926',
    '927',
    '928',
    '929',
    '930',
    '931',
    '932',
    '933',
    '934',
    '935',
    '936',
    '937',
    '938',
    '939',
    '940',
    '941',
    '942',
    '943',
    '944',
    '945',
    '946',
    '947',
    '948',
    '949',
    '950',
    '951',
    '952',
    '953',
    '954',
    '955',
    '956',
    '957',
    '958',
    '959',
    '960',
    '961',
    '962',
    '963',
    '964',
    '965',
    '966',
    '967',
    '968',
    '969',
    '970',
    '971',
    '972',
    '973',
    '974',
    '975',
    '976',
    '977',
    '978',
    '979',
    '980',
    '981',
    '982',
    '983',
    '984',
    '985',
    '986',
    '987',
    '988',
    '989',
    '990',
    '991',
    '992',
    '993',
    '994',
    '995',
    '996',
    '997',
    '998',
    '999',
    '1000'
  ])
};

Button.defaultProps = {
  action: 'foo'
};
